e2etest/declarativeHelpers.go

```diff
@@ -0,0 +1,528 @@
+// Copyright © Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package e2etest
+
+import (
+	"errors"
+	"fmt"
+	"github.com/Azure/azure-storage-azcopy/cmd"
+	"github.com/Azure/azure-storage-azcopy/common"
+	"github.com/JeffreyRichter/enum/enum"
+	"math"
+	"path"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+///////////
+
+// folder is syntactic sugar to make folders stand out in our file lists
+func folder(s string) string {
+	return strings.TrimRight(s, "/") + "/" // ensure it ends in one slash
+}
+
+func isFolder(s string) bool {
+	return strings.HasSuffix(s, "/")
+}
+
+func isRootFolder(s string) bool {
+	return s == "/"
+}
+
+func asFolderName(s string) string {
+	return strings.TrimRight(s, "/") // strip the trialing /
+}
+
+func asFolderDummyContent(s string) string { // make a dummy filename in the folder
+	return path.Join(s, "dummy")
+}
+
+///////////
+
+var sanitizer = common.NewAzCopyLogSanitizer() // while this is "only tests", we may as well follow good SAS redaction practices
+
+////////
+
+type comparison struct {
+	equals bool
+}
+
+func (c comparison) String() string {
+	if c.equals {
+		return "equals"
+	} else {
+		return "notEquals"
+	}
+}
+
+func equals() comparison {
+	return comparison{true}
+}
+
+func notEquals() comparison {
+	return comparison{false}
+}
+
+///////
+
+// simplified assertion interface. Allows us to abstract away the specific test harness we are using
+// (in case we change it... again)
+type asserter interface {
+	Assert(obtained interface{}, comp comparison, expected interface{}, comment ...string)
+	AssertNoErr(err error, comment ...string)
+	Error(reason string)
+	Skip(reason string)
+	Failed() bool
+
+	// ScenarioName piggy-backs on this interface, in a context-value-like way (ugly, but it works)
+	CompactScenarioName() string
+}
+
```

Comment:
I would suggest modeling our assertion somewhat around Gocheck's assertion interface.

This would allow for custom comparisons (and for us to log a diff or something else of that nature).

Reply:
It is somewhat similar to that interface, with two notable differences that come to mind:
1. No checker object.  The rationale for that is that (a) as soon as we commit to a particular checker object, that ties us (somewhat) to a particular framework, which is the thing this interface was added to avoid.  And (b), it sees like we don't need to actually write many assertions. E.g. the typical test, created with this framework, contains no assertions at all in the test itself. (And that's consistent with the goal of a declarative style).  For this reason, I feel it's not clear that more advanced comparisons would give us much benefit.
2. Our assert method does not cause the executing test to stop. I.e. our "Assert" == their "Check".  Maybe we should rename ours, or maybe not. (AFAICT Assert doesn't mean AssertAndStopIfFailed in other Go test frameworks)  In any case , I think it's correct that our main assertion/check method does allow the test to carry on. So you can see, for instance, information about all files that were transferred wrong, instead of just one.  Or, if the lengths of two lists don't match, we don't stop at the length check, but instead also output more info about the individual differences. (e.g. specific files that aren't found)

I'm going to leave it as-is for now, but you could easily do renames later.